### ISSUE TITLE
[Bug]: read_remote not working on WIN ?

### ISSUE THREAD

**Issue by atravert** [spectrochempy/spectrochempy#622]
### What happened?

scp.read_remote("http://www.eigenvector.com/data/Corn/corn.mat")

returns None and the following warning:
` WARNING | (UserWarning) [Errno 22] Invalid argument: 'http://www.eigenvector.com/data/Corn/corn.mat'`

This is because in line 215 of importer.py, `read_(self.objtype(), filename, **kwargs)` returns an OSError.

The following seems to fix it:

```
            try:
                res = read_(self.objtype(), filename, **kwargs)
                # sometimes read_ can return None (e.g. non labspec text file)

            except OSError as e:
                # in windows, passing an url gives [Errno 22] Invalid argument
                if e.args[0] == 22:
                    kwargs["read_method"] = read_
                    res = _read_remote(self.objtype(), filename, **kwargs)
                else:
                    raise e

            except FileNotFoundError:
                # try to get the file from github or url
                kwargs["read_method"] = read_
                try:
                    res = _read_remote(self.objtype(), filename, **kwargs)
```

### Environment

INSTALLED VERSIONS
------------------
commit: aae963a1caa6023cb244fe4e07e1cfe6551a53e1
python: 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:00:38) [MSC v.1934 64 bit (AMD64)]
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 12, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: ('en_US', 'ISO8859-1')
quadprog: 0.1.11
brukeropusreader: 1.3
numpy-quaternion: None
cffconvert: 2.0.0
colorama: 0.4.6
dill: 0.3.6
docrep: 0.3.2
ipython: None
jinja2: 3.1.2
matplotlib: 3.7.1
numba: 0.56.4
numpy: 1.23.5
numpydoc: 1.5.0
pint: 0.20.1
pytz: 2023.3
requests: 2.28.2
scipy: 1.10.1
tqdm: 4.65.0
traitlets: 5.9.0
scikit-learn: None
xlrd: 2.0.1
pyyaml: None
ipywidgets==8.0.4: None
widgetsnbextension==4.0.5: None
jupyterlab_widgets==3.0.5: None
ipympl: 0.9.3
setuptools: 67.6.1
setuptools_scm: 7.1.0
gitpython: None
jupyterlab: 3.5.3
nodejs: 0.1.1
pytest: 7.2.2
pytest-doctestplus: None
pytest-flake8: None
pytest-mock: None
pyfakefs: 5.2.0
pep8-naming: None
scikit-image: None
coverage: 7.2.2
black: 23.3.0
pre-commit: None
mamba: 1.4.1
docrep: 0.3.2
 WARNING | (UserWarning) Sphinx Gallery in version 0.12.2 is not supported by Jupytext. Please use sphinx-gallery<=0.7.0 instead. If that is an issue, feel free to report it at https://github.com/mwouts/jupytext/issues, or even better, prepare a PR to handle the new signature of sphinx_gallery.notebook.rst2md.
--- Logging error ---
Traceback (most recent call last):
  File "C:\Miniconda3\envs\scpy-a\lib\logging\handlers.py", line 74, in emit
    self.doRollover()
  File "C:\Miniconda3\envs\scpy-a\lib\logging\handlers.py", line 179, in doRollover
    self.rotate(self.baseFilename, dfn)
  File "C:\Miniconda3\envs\scpy-a\lib\logging\handlers.py", line 115, in rotate
    os.rename(source, dest)
PermissionError: [WinError 32] Le processus ne peut pas accéder au fichier car ce fichier est utilisé par un autre processus: 'C:\\Users\\traverta\\.spectrochempy\\logs\\spectrochempy.log' -> 'C:\\Users\\traverta\\.spectrochempy\\logs\\spectrochempy.log.1'
Call stack:
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\pydevconsole.py", line 568, in <module>
    pydevconsole.start_client(host, port)
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\pydevconsole.py", line 496, in start_client
    process_exec_queue(interpreter)
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\pydevconsole.py", line 287, in process_exec_queue
    interpreter.add_exec(code_fragment)
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_code_executor.py", line 109, in add_exec
    more, exception_occurred = self.do_add_exec(code_fragment)
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_ipython_console.py", line 36, in do_add_exec
    more, exception_occurred = self.interpreter.add_exec(code_fragment.text)
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_ipython_console_011.py", line 382, in add_exec
    res = self.ipython.run_cell(line, store_history=True)
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\IPython\core\interactiveshell.py", line 3006, in run_cell
    result = self._run_cell(
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\IPython\core\interactiveshell.py", line 3061, in _run_cell
    result = runner(coro)
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\IPython\core\async_helpers.py", line 129, in _pseudo_sync_runner
    coro.send(None)
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\IPython\core\interactiveshell.py", line 3266, in run_cell_async
    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\IPython\core\interactiveshell.py", line 3445, in run_ast_nodes
    if await self.run_code(code, result, async_=asy):
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\IPython\core\interactiveshell.py", line 3505, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-41f652028d13>", line 1, in <module>
    scp.show_versions()
  File "C:\Users\traverta\Dropbox\Programmes\spectrochempy\spectrochempy\utils\print_versions.py", line 104, in show_versions
    mod = optional.import_optional_dependency(dep, errors="ignore")
  File "C:\Users\traverta\Dropbox\Programmes\spectrochempy\spectrochempy\utils\optional.py", line 96, in import_optional_dependency
    module = importlib.import_module(name)
  File "C:\Miniconda3\envs\scpy-a\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\jupytext\__init__.py", line 3, in <module>
    from .formats import NOTEBOOK_EXTENSIONS, get_format_implementation, guess_format
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\jupytext\formats.py", line 13, in <module>
    from .cell_reader import (
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Miniconda3\envs\scpy-a\lib\site-packages\jupytext\cell_reader.py", line 20, in <module>
    warnings.warn(
  File "C:\Miniconda3\envs\scpy-a\lib\warnings.py", line 109, in _showwarnmsg
    sw(msg.message, msg.category, msg.filename, msg.lineno,
  File "C:\Users\traverta\Dropbox\Programmes\spectrochempy\spectrochempy\application\application.py", line 816, in _custom_warning
    self.log.warning(f"({category.__name__}) {message}")
Message: '(UserWarning) Sphinx Gallery in version 0.12.2 is not supported by Jupytext. Please use sphinx-gallery<=0.7.0 instead. If that is an issue, feel free to report it at https://github.com/mwouts/jupytext/issues, or even better, prepare a PR to handle the new signature of sphinx_gallery.notebook.rst2md.'
Arguments: ()
jupytext: 1.14.5
sphinx==5.3: None
sphinx_rtd_theme==1.2: None
autodocsumm: 0.2.6
sphinx-gallery: None
nbsphinx: 0.9.1
jupyter_sphinx: 0.4.0
json5: 0.9.5
sphinx-copybutton: None
sphinxcontrib-bibtex: None
pandoc==2.19: None

**Comment by fernandezc** [comment]
@travert The error message is not very informative because it is shorten for the end user. For debugging you should comment lines 982 to 987 in application/application.py.  or eventually set the environment variable SCPY_TESTING to True/1. 
So you will more easily locate the source of the problem.

**Comment by fernandezc** [comment]
I mean we need to know why there is a OSerror in line 215

**Comment by fernandezc** [comment]
> @travert The error message is not very informative because it is shorten for the end user. For debugging you should comment lines 982 to 987 in application/application.py. or eventually set the environment variable SCPY_TESTING to True/1.
> So you will more easily locate the source of the problem.

Actually the test for catching or not exception was wrong I modified it in upstream/master. Full message will appear if you are in purest (hopefully!) but not in normal operation unless you set the environment variable.

**Comment by atravert** [comment]
The error comes from _read_mat() line 83:
`fid = open(filename, "rb")`

_read_mat() is called in line 215 of importer.oy : `res = read_(self.objtype(), filename, **kwargs)` 

where :
- self.objtype() is an empty NDDataset
- filename is the url
- kwargs are:  {'remote': True, 'merge': False, 'replace_existing': False, 'read_only': True}

so i understand why the error is raised -- the url is indeed an invalid argument of open() -- but I don't figure out how it should work normally... (I don't have a ubuntu os here)

**Comment by fernandezc** [comment]
ok  then:
When the method read_ (_read_mat) is called line 215,  in in read_mat 
line 83:   fid = open(filename, "rb")
generate an error because the file is not found. On my computer FileNotFound is the error type, but if an OSError is generated on windows, we could just write on line 217: 

      except (FileNotFoundError, OSError): 

instead of only 

      except FileNotFoundError:

at

**Comment by atravert** [comment]
tnhanks ! I will try.  I have tried a similar fix (catching OSError), but this led read_remote() not working anymore with spectrochempy_data remote files, on both win and ubuntu tests... Let me try it, I will let you know :-)

**Comment by fernandezc** [comment]
I think I fixed the errors, but I can't test with window in workflow. You have to do it locally for now. I will look at the workflow problem tonight.

**Comment by fernandezc** [comment]
> I think I fixed the errors, but I can't test with window in workflow. You have to do it locally for now. I will look at the workflow problem tonight.

@atravert  Do the latest changes solve the problem on WIN?

**Comment by atravert** [comment]
Locally, on my PC, It works for url but not for missing spectrochempy_data file (returns a file not found error). This could be related to a previous change, but as my connection is not very good It is difficult to determine it more precisely....

**Comment by atravert** [comment]
I found it: the error comes from importer.py line 676:   
 url = (
            f"https://github.com/spectrochempy/spectrochempy_data/raw/master/"
            f"testdata/{url}"
        )

on my PC this returns an url with \\ , for instance:

"https://github.com/spectrochempy/spectrochempy_data/raw/master/testdata\\irdata/nh4-activation.spg" 

it is foxed by using url.as_posix():

        url = (
            f"https://github.com/spectrochempy/spectrochempy_data/raw/master/"
            f"testdata/{url.as_posix()}"
        )

**Comment by fernandezc** [comment]
Should be resolved by PR #628.
